Firestore/core/src/index/firestore_index_value_writer.cc

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "Firestore/core/src/index/firestore_index_value_writer.h"
+
+#include <cmath>
+#include <limits>
+#include <string>
+
+#include "Firestore/core/src/model/resource_path.h"
+#include "Firestore/core/src/nanopb/nanopb_util.h"
+
+namespace firebase {
+namespace firestore {
+namespace index {
+namespace {
+
+// Note: This code is copied from the backend. Code that is not used by
+// Firestore was removed.
+
+// The client SDK only supports references to documents from the same database.
+// We can skip the first five segments.
+constexpr int DocumentNameOffset = 5;
+
+enum IndexType {
+  kNull = 5,
+  kBoolean = 10,
+  kNan = 13,
+  kNumber = 15,
+  kTimestamp = 20,
+  kString = 25,
+  kBlob = 30,
+  kReference = 37,
+  kGeopoint = 45,
+  kArray = 50,
+  kMap = 55,
+  kReferenceSegment = 60,
+  // A terminator that indicates that a truncatable value was not truncated.
+  // This must be smaller than all other type labels.
+  kNotTruncated = 2
+};
+
+// The write methods below short-circuit writing terminators for values
+// containing a (terminating) truncated value.
+//
+// As an example, consider the resulting encoding for:
+//
+// ["bar", [2, "foo"]] -> (STRING, "bar", TERM, ARRAY, NUMBER, 2, STRING, "foo",
+// TERM, TERM, TERM)
+// ["bar", [2, truncated("foo")]] -> (STRING, "bar", TERM, ARRAY, NUMBER, 2,
+// STRING, "foo", TRUNC)
+// ["bar", truncated(["foo"])] -> (STRING, "bar", TERM, ARRAY. STRING, "foo",
+// TERM, TRUNC)
```

Comment:
I think this is actually not relevant to our SDK since we don't support truncation. Will send a PR to remove from Android.

Reply:
Removed.